csfieldguide/chapters/content/en/coding-compression/sections/general-purpose.md

```diff
@@ -43,27 +31,337 @@ Here are some of the bits from the example earlier in this chapter; you can past
+| 8    | 4    | IA     | 10                  | MI                | IA?                         |
+| 9    | 2    | M      | 11                  | IAM               | M?                          |
+
+Once we have decoded the whole message we now have the following string:
+
+```
+IAMSAMSAMIAM
+```
+
+So we successfully encoded and decoded the message!
+And this is the dictionary that the decoder created:
+
+| Code | Bits         | Sequence |
+|------|--------------|----------|
+| 0    | 000000000000 | I        |
+| 1    | 000000000001 | A        |
+| 2    | 000000000010 | M        |
+| 3    | 000000000011 | S        |
+| 4    | 000000000100 | IA       |
+| 5    | 000000000101 | AM       |
+| 6    | 000000000110 | MS       |
+| 7    | 000000000111 | SA       |
+| 8    | 000000001000 | AMS      |
+| 9    | 000000001001 | SAM      |
+| 10   | 000000001010 | MI       |
+| 11   | 000000001011 | IAM      |
+
+Notice that it is exactly the same dictionary as the encoder created - the dictionary has been reconstructed without having to send it to the decoder!
+
+
+{panel type="extra-for-experts"}
+
+# Other languages
 
-The [CS Unplugged site has activities and information about the Ziv-Lempel approach](http://csunplugged.org/text-compression),
-and the ["Computing Science Inside" site also has an activity based on this method](https://web.archive.org/web/20150311225517/http://csi.dcs.gla.ac.uk/workshop-view.php?workshopID=1).
-The CS4FN site discusses [a related approach which is a little simpler, but not so useful in practice](http://www.cs4fn.org/internet/crushed.php).
+But what about different languages? Since the dictionary builds up based on the characters it has seen, it will simply adapt to whatever language is being used, whether it is English, French, Chinese, or even a computer programming language.
+Each language will have its own common substrings, and these will end up in the dictionary.
 
 {panel end}
 
-{panel type="curiosity"}
+{panel type="extra-for-experts"}
 
-# ZL or LZ compression?
+# Representing the codes in fewer than 12 bits
 
-The method we have described here is named "Ziv-Lempel" compression after Jacob Ziv and Abraham Lempel, the two computer scientists who invented it in the 1970s.
-Unfortunately someone mixed up the order of their names when they wrote an article about it, and called it "LZ" compression instead of "ZL" compression.
-So many people copied the mistake that Ziv and Lempel’s method is now usually called "LZ compression"!
+While the dictionary is being built initially it doesn’t make sense to use 12 bits to store so few entries.
+Instead, while there are fewer than 4 entries we could use 2 bits, when there are more than 4 but fewer than 8 entries we could used 3 bits, etc.
+Eventually the dictionary will fill all 4096 (12 bit) entries, then what do we do? There are different ways to deal with this: a simple system just keeps the dictionary the same from then on and hopes the nature of the text doesn’t change.
```

Comment:
New sentence on a new line - search for `period-space`, `questionmark-space` etc to find others. Though if the sentence is short enough (a word or two) you don't need to bother